File: src/moveit_r2_config/src/move_hand_node.cpp
#include <functional>
#include <memory>
#include <string>

#include <geometry_msgs/msg/vector3.hpp>
#include <moveit/move_group_interface/move_group_interface.h>
#include <rclcpp/rclcpp.hpp>

class MoveHandController
{
public:
  explicit MoveHandController(const rclcpp::Node::SharedPtr & node)
  : node_(node)
  {
    group_name_ = node_->declare_parameter<std::string>("group_name", "arm_with_hand");
    tip_link_ = node_->declare_parameter<std::string>("tip_link", "link5_1");
    topic_name_ = node_->declare_parameter<std::string>("topic_name", "/move_hand");
    position_tolerance_ = node_->declare_parameter<double>("position_tolerance", 0.005);
    orientation_tolerance_ = node_->declare_parameter<double>("orientation_tolerance", 0.05);
    planning_time_ = node_->declare_parameter<double>("planning_time", 5.0);
    planning_attempts_ = node_->declare_parameter<int>("planning_attempts", 5);

    move_group_ = std::make_shared<moveit::planning_interface::MoveGroupInterface>(node_, group_name_);
    move_group_->setEndEffectorLink(tip_link_);
    move_group_->setGoalPositionTolerance(position_tolerance_);
    move_group_->setGoalOrientationTolerance(orientation_tolerance_);
    move_group_->setPlanningTime(planning_time_);
    move_group_->setNumPlanningAttempts(planning_attempts_);

    using std::placeholders::_1;
    subscription_ = node_->create_subscription<geometry_msgs::msg::Vector3>(
      topic_name_, rclcpp::QoS(10),
      std::bind(&MoveHandController::moveHandCallback, this, _1));

    RCLCPP_INFO(
      node_->get_logger(), "Listening to %s for relative commands on %s",
      topic_name_.c_str(), tip_link_.c_str());
  }

private:
  void moveHandCallback(const geometry_msgs::msg::Vector3::SharedPtr msg)
  {
    auto current_pose = move_group_->getCurrentPose(tip_link_);
    auto target_pose = current_pose.pose;

    // Vector3 offsets are interpreted directly in the world/base frame.
    target_pose.position.x += msg->x;
    target_pose.position.y += msg->y;
    target_pose.position.z += msg->z;

    move_group_->setStartStateToCurrentState();
    move_group_->setPoseTarget(target_pose, tip_link_);
    auto const result = move_group_->move();
    move_group_->stop();
    move_group_->clearPoseTargets();

    if (result == moveit::core::MoveItErrorCode::SUCCESS) {
      RCLCPP_INFO(
        node_->get_logger(), "Moved %s by Vector3(%.3f, %.3f, %.3f)",
        tip_link_.c_str(), msg->x, msg->y, msg->z);
    } else {
      RCLCPP_WARN(
        node_->get_logger(),
        "MoveIt planning/execution failed for Vector3(%.3f, %.3f, %.3f) (error code: %d)",
        msg->x, msg->y, msg->z, result.val);
    }
  }

  rclcpp::Node::SharedPtr node_;
  std::string group_name_;
  std::string tip_link_;
  std::string topic_name_;
  double position_tolerance_;
  double orientation_tolerance_;
  double planning_time_;
  int planning_attempts_;
  std::shared_ptr<moveit::planning_interface::MoveGroupInterface> move_group_;
  rclcpp::Subscription<geometry_msgs::msg::Vector3>::SharedPtr subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<rclcpp::Node>("move_hand_node");
  MoveHandController controller(node);
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: install/moveit_r2_config/lib/moveit_r2_config/move_hand_node.py
#!/usr/bin/env python3

import sys
from typing import Optional

import moveit_commander
import rclpy
from geometry_msgs.msg import Pose, Vector3
from moveit_commander import MoveGroupCommander
from rclpy.node import Node
from tf_transformations import quaternion_matrix


class MoveHandNode(Node):
    """Apply relative end-effector shifts received on /move_hand."""

    def __init__(self) -> None:
        super().__init__("move_hand_node")

        self.group_name: str = (
            self.declare_parameter("group_name", "arm_with_hand").get_parameter_value().string_value
        )
        self.tip_link: str = (
            self.declare_parameter("tip_link", "link5_1").get_parameter_value().string_value
        )
        topic_name: str = (
            self.declare_parameter("topic_name", "/move_hand").get_parameter_value().string_value
        )

        self.move_group: Optional[MoveGroupCommander] = None
        try:
            self.move_group = MoveGroupCommander(self.group_name)
        except RuntimeError as err:
            self.get_logger().error(f"MoveGroupCommander init failed: {err}")
            raise

        self.move_group.set_end_effector_link(self.tip_link)

        self.create_subscription(Vector3, topic_name, self._move_hand_callback, 10)
        self.get_logger().info(
            f"Listening for Vector3 commands on {topic_name} to move {self.tip_link} of {self.group_name}"
        )

    def destroy_node(self) -> bool:
        if self.move_group:
            self.move_group.stop()
            self.move_group.clear_pose_targets()
        return super().destroy_node()

    def _move_hand_callback(self, msg: Vector3) -> None:
        if not self.move_group:
            self.get_logger().error("Move group is not initialized.")
            return

        current_pose = self.move_group.get_current_pose(self.tip_link).pose
        target_pose = Pose()
        target_pose.orientation = current_pose.orientation

        relative = [msg.x, msg.y, msg.z]
        quaternion = [
            current_pose.orientation.x,
            current_pose.orientation.y,
            current_pose.orientation.z,
            current_pose.orientation.w,
        ]
        rotation_matrix = quaternion_matrix(quaternion)[:3, :3]
        world_delta = rotation_matrix.dot(relative)

        target_pose.position.x = current_pose.position.x + world_delta[0]
        target_pose.position.y = current_pose.position.y + world_delta[1]
        target_pose.position.z = current_pose.position.z + world_delta[2]

        self.move_group.set_pose_target(target_pose, self.tip_link)
        success = self.move_group.go(wait=True)
        self.move_group.stop()
        self.move_group.clear_pose_targets()

        if success:
            self.get_logger().info(
                f"Moved {self.tip_link} by relative Vector3({msg.x:.3f}, {msg.y:.3f}, {msg.z:.3f})"
            )
        else:
            self.get_logger().warn("MoveIt failed to plan/execute the requested motion.")


def main(args=None) -> None:
    rclpy.init(args=args)
    moveit_commander.roscpp_initialize(sys.argv)
    node = MoveHandNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        moveit_commander.roscpp_shutdown()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
